### PR TITLE
Expand scheduling configurer so AppInsights can register full journey of dependencies

### PR DESCRIPTION
**Pull request by doncem**
### Change description ###

Finally coming to another service. Backporting same config as it is in hmcts/send-letter-service. Proven longly as working. Further AppInsights improvements to come. Inter-remote-hidden communication between processor and orchestrator is the goal

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/config/SchedulerConfig.java [hunk 1]
```diff
@@ -1,21 +1,129 @@
...
+        private RequestTelemetryContext requestContext;
+
+        WrappedRunnable(Runnable task, RequestTelemetryContext requestContext) {
+            this.task = task;
+            this.requestContext = requestContext;
+        }
+
+        @Override
+        public void run() {
+            if (ThreadContext.getRequestTelemetryContext() != null) {
```
Comment (apereverzin) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/config/SchedulerConfig.java]: What if '== null'?

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/config/SchedulerConfig.java]: `requestContext` will be just set as in line 124.

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/config/SchedulerConfig.java]: Ah, I see

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/config/SchedulerConfig.java]: 'requestContext' class attribute does not look necessary, it is refreshed each time and set to ThreadContext. It looks it is possible just do this:

RequestTelemetryContext requestContext = REQUEST_CONTEXT_SUPPLIER.get();
ThreadContext.setRequestTelemetryContext(requestContext);

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/config/SchedulerConfig.java]: This is following Microsoft's application insights guidelines on how to track schedules correctly: https://github.com/microsoft/ApplicationInsights-Java/wiki/Distributed-Tracing-in-Asynchronous-Java-Applications#context-propagation-in-scheduled-events.

I believe `ThreadContext` can be cleared at any point in the request lifetime. Most likely once it's published via telemetry client. Then new registered context will be used upon next schedule run making this `if` skipped

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/config/SchedulerConfig.java]: I see, this is just how Microsoft recommends

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/config/SchedulerConfig.java [hunk 2]
```diff
@@ -1,21 +1,129 @@
...
+
+        @Override
+        public void execute(Runnable command) {
+            super.execute(new WrappedRunnable(command, REQUEST_CONTEXT_SUPPLIER.get()));
+        }
+
+        @Override
+        public ScheduledFuture<?> schedule(Runnable task, Trigger trigger) {
+            return super.schedule(new WrappedRunnable(task, REQUEST_CONTEXT_SUPPLIER.get()), trigger);
+        }
```
Comment (apereverzin) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/config/SchedulerConfig.java]: Makes sense to move call to WrappedRunnable constructor to a method.

**Comment by doncem** [comment]
Letting current projects to settle in. Even though this is the state of working wrapper proven in another project, but.. as I said - letting other projects to settle in

**Pull request merged**